# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/clients/manager/ClientManagerService.h
# Commit: 6125bb02952a
# Full Hash: 6125bb02952aff71c0d757e238a7752101b2d6c2
# Author: Jean-Yves Avenard <jyavenard@mozilla.com>
# Date: 2019-09-03 04:06:20
# Regressor Bug: 1556489
# File Overlap Count: 1
# Description:
#   Bug 1556489 - P23 - Only create new ClientSource objects in the content, but use the recreated info from redirects in the parent. r=asuth
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D42260
# ==============================================================================

diff -r 06529d12b589 -r 6125bb02952a dom/clients/manager/ClientManagerService.h
--- a/dom/clients/manager/ClientManagerService.h	Mon Aug 26 12:18:57 2019 +1000
+++ b/dom/clients/manager/ClientManagerService.h	Tue Aug 27 08:53:37 2019 +1000
@@ -21,6 +21,7 @@
 
 class ClientManagerParent;
 class ClientSourceParent;
+class ClientHandleParent;
 class ContentParent;
 
 // Define a singleton service to manage client activity throughout the
@@ -31,6 +32,10 @@
   // optimize for insertion, removal, and lookup by UUID.
   nsDataHashtable<nsIDHashKey, ClientSourceParent*> mSourceTable;
 
+  // The set of handles waiting for their corresponding ClientSourceParent
+  // to be created.
+  nsDataHashtable<nsIDHashKey, nsTArray<ClientHandleParent*>> mPendingHandles;
+
   nsTArray<ClientManagerParent*> mManagerList;
 
   bool mShutdown;
@@ -53,6 +58,12 @@
   ClientSourceParent* FindSource(
       const nsID& aID, const mozilla::ipc::PrincipalInfo& aPrincipalInfo);
 
+  // Called when a ClientHandle is created before the corresponding
+  // ClientSource. Will call FoundSource on the ClientHandleParent when it
+  // becomes available.
+  void WaitForSource(ClientHandleParent* aHandle, const nsID& aID);
+  void StopWaitingForSource(ClientHandleParent* aHandle, const nsID& aID);
+
   void AddManager(ClientManagerParent* aManager);
 
   void RemoveManager(ClientManagerParent* aManager);